projects/miopen/test/gtest/nonstandardsquash_tensor_op.cpp

```diff
@@ -0,0 +1,337 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <miopen/tensor_ops.hpp>
+#include <tensor_util.hpp>
+#include "gtest_common.hpp"
+
+namespace {
+std::vector<std::vector<size_t>> tensorALensArr    = {{1, 1, 4}, {1, 1, 8}, {1, 1, 32}}; // tensor A
+std::vector<std::vector<size_t>> tensorAStridesArr = {{4, 4, 1}, {8, 8, 1}, {32, 32, 1}};
+
+std::vector<std::vector<size_t>> tensorBLensArr = {{1, 16, 4}, {1, 32, 8}, {1, 8, 32}}; // tensor B
+std::vector<std::vector<size_t>> tensorBStridesArr = {
+    {16 * 4, 4 * 1, 1}, {32 * 8, 8 * 1, 1}, {8 * 32, 32 * 1, 1}};
+
+std::vector<std::vector<int64_t>> offsetsArr = {
+    {0, 0, 0}, {64, 32, 16}, {32, 16, 32}, {32, 16, 32}};
+
+std::vector<std::vector<float>> alphabetaArr = {{1, 1, 0}, {-1, 1, 1}, {1.0, 0.5, 0.3}};
+
+std::vector<bool> packedArr = {true, false};
+
+std::vector<miopenTensorOp_t> operationArr = {
+    miopenTensorOpAdd, miopenTensorOpMul, miopenTensorOpMin, miopenTensorOpMax};
+} // namespace
+
+struct TestCase
+{
+    std::vector<size_t> tensorlens_ac;
+    std::vector<size_t> tensorlens_b;
+    std::vector<int64_t> offsets;
+    std::vector<size_t> stride_a;
+    std::vector<size_t> stride_b;
+    std::vector<size_t> stride_c;
+    std::vector<float> alphabeta;
+    bool packed;
+    miopenTensorOp_t operation;
+};
+
+template <typename T>
+struct TensorOpsCommon : public testing::TestWithParam<TestCase>
+{
+    void SetUp() override { prng::reset_seed(); }
+
+    void Run()
+    {
+        CreateTensors();
+
+        std::vector<T> tensorGPUData = CalculateOnGPU();
+        std::vector<T> tensorCPUData = CalculateOnCPU();
+
+        CompareResults(tensorGPUData, tensorCPUData);
+    }
+
+private:
+    void CreateTensors()
+    {
+        const TestCase& testCase = GetParam();
+
+        tensorA = CreateTensor(
+            testCase.tensorlens_ac, testCase.stride_a, testCase.offsets[0], testCase.packed);
+        tensorB = CreateTensor(
+            testCase.tensorlens_b, testCase.stride_b, testCase.offsets[1], testCase.packed);
+        tensorC = CreateTensor(
+            testCase.tensorlens_ac, testCase.stride_c, testCase.offsets[2], testCase.packed);
+    }
+
+    tensor<T> CreateTensor(const std::vector<size_t>& lens,
+                           const std::vector<size_t>& strides,
+                           int64_t offset,
+                           bool isPacked)
+    {
+        uint64_t max_value = miopen_type<T>{} == miopenHalf ? 5 : 17;
+
+        if(!isPacked)
+        {
+            std::vector<size_t> real_strides(strides.begin() + (strides.size() - lens.size()),
+                                             strides.end());
+            auto r = tensor<T>{lens, real_strides}.generate(tensor_elem_gen_integer{max_value});
+            r.data.resize(r.data.size() + offset);
+            return r;
```

Comment:
I believe this will add offset to the end of the tensor data, but below it looks like we are accessing the values offset from the start. I think we want to have empty values to start with, and then add the generated values to the end. Or maybe just generate extra values for the whole block including offsets. The way it is now, I think we will run on a bunch of default constructed T.

For example:
Values are {0, 1, 2}, offset = 2, data-type int.
After resize I believe it'll be {0, 1, 2, 0, 0}.
The first a_idx would be 2, and I think it would operate over {2, 0, 0}, but I would expect we want {0, 1, 2}. 

Below in CalculateOnCPU:
```
        auto a_idx = testCase.offsets[0];
        auto b_idx = testCase.offsets[1];
        auto c_idx = testCase.offsets[2];

        // Currently non standard squashed operation is supported only on 3D tensors
        // A/C tensors are 1x1xH, tensor B is 1xCxH (C > 1)
        // This is simulated 3D non standard squashed tensor operation
        for(auto i = 0; i < testCase.tensorlens_ac[2]; i++)
        {
            auto a_val = a.data[a_idx] * testCase.alphabeta[0];
            auto c_val = c.data[c_idx] * testCase.alphabeta[2];
            for(auto j = 0; j < testCase.tensorlens_b[1]; j++)
            {
                c_val += dataOp(a_val, b.data[b_idx]);
                b_idx += testCase.stride_b[1];
            }
            c.data[c_idx] = c_val;
            a_idx += testCase.stride_a[2];
            c_idx += testCase.stride_c[2];
            b_idx = testCase.offsets[1] + (i + 1) * testCase.stride_b[2];
        }
```

Reply:
I agree, this seems to be an incorrectness coming from the old ctest. This is also in the `ternary_tensor_ops` test, will update it both. Also, offset values should be added in the `else` branch below.

I think we should just prepend values to the tensor
```suggestion
            auto r = tensor<T>{lens, real_strides}.generate(tensor_elem_gen_integer{max_value});
            r.data.insert(r.data.begin(), offset, T());
            return r;
```